// @dedot/contracts/cjs/Contract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Contract = void 0;
const codecs_1 = require("@dedot/codecs");
const TypinkRegistry_js_1 = require("./TypinkRegistry.js");
const index_js_1 = require("./executor/index.js");
const utils_js_1 = require("./utils.js");
class Contract {
    client;
    #registry;
    #address;
    #metadata;
    #options;
    constructor(client, metadata, address, options) {
        this.client = client;
        (0, utils_js_1.ensureSupportContractsPallet)(client);
        this.#address = new codecs_1.AccountId32(address);
        this.#metadata = typeof metadata === 'string' ? (0, utils_js_1.parseRawMetadata)(metadata) : metadata;
        this.#registry = new TypinkRegistry_js_1.TypinkRegistry(this.#metadata);
        this.#options = options;
    }
    decodeEvent(eventRecord) {
        return this.#registry.decodeEvent(eventRecord, this.address);
    }
    decodeEvents(eventRecords) {
        return this.#registry.decodeEvents(eventRecords, this.address);
    }
    get metadata() {
        return this.#metadata;
    }
    get address() {
        return this.#address;
    }
    get registry() {
        return this.#registry;
    }
    get query() {
        return (0, utils_js_1.newProxyChain)(new index_js_1.QueryExecutor(this.client, this.#registry, this.#address, this.#options));
    }
    get tx() {
        return (0, utils_js_1.newProxyChain)(new index_js_1.TxExecutor(this.client, this.#registry, this.#address, this.#options));
    }
    get events() {
        return (0, utils_js_1.newProxyChain)(new index_js_1.EventExecutor(this.client, this.#registry, this.#address, this.#options));
    }
    get options() {
        return this.#options;
    }
}
exports.Contract = Contract;
