// @dedot/contracts/cjs/ContractDeployer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractDeployer = void 0;
const TypinkRegistry_js_1 = require("./TypinkRegistry.js");
const ConstructorQueryExecutor_js_1 = require("./executor/ConstructorQueryExecutor.js");
const index_js_1 = require("./executor/index.js");
const utils_js_1 = require("./utils.js");
class ContractDeployer {
    client;
    #metadata;
    #registry;
    #code;
    #options;
    constructor(client, metadata, codeHashOrWasm, options) {
        this.client = client;
        (0, utils_js_1.ensureSupportContractsPallet)(client);
        this.#metadata = typeof metadata === 'string' ? (0, utils_js_1.parseRawMetadata)(metadata) : metadata;
        this.#registry = new TypinkRegistry_js_1.TypinkRegistry(this.#metadata);
        this.#code = codeHashOrWasm;
        this.#options = options;
    }
    get metadata() {
        return this.#metadata;
    }
    get registry() {
        return this.#registry;
    }
    get tx() {
        return (0, utils_js_1.newProxyChain)(new index_js_1.ConstructorTxExecutor(this.client, this.#registry, this.#code, this.#options));
    }
    get query() {
        return (0, utils_js_1.newProxyChain)(new ConstructorQueryExecutor_js_1.ConstructorQueryExecutor(this.client, this.#registry, this.#code, this.#options));
    }
    get options() {
        return this.#options;
    }
}
exports.ContractDeployer = ContractDeployer;
