// @dedot/codecs/cjs/codecs/generic/AccountId32.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$AccountId32 = exports.AccountId32 = exports.accountId32ToHex = void 0;
const $ = __importStar(require("@dedot/shape"));
const utils_1 = require("@dedot/utils");
const accountId32ToHex = (input) => {
    if (input instanceof AccountId32) {
        return input.raw;
    }
    else if ((0, utils_1.isU8a)(input)) {
        return (0, utils_1.u8aToHex)(input);
    }
    else if ((0, utils_1.isString)(input)) {
        return (0, utils_1.u8aToHex)((0, utils_1.decodeAddress)(input));
    }
    else if ((0, utils_1.isHex)(input)) {
        return input;
    }
    throw Error(`Invalid input for AccountId32: ${input}`);
};
exports.accountId32ToHex = accountId32ToHex;
class AccountId32 {
    raw;
    constructor(input) {
        this.raw = (0, exports.accountId32ToHex)(input);
    }
    address(ss58Format) {
        return (0, utils_1.encodeAddress)(this.raw, ss58Format);
    }
    toJSON() {
        return this.address();
    }
    eq(other) {
        return this.raw === new AccountId32(other).raw;
    }
}
exports.AccountId32 = AccountId32;
exports.$AccountId32 = $.instance(AccountId32, $.Tuple($.FixedHex(32)), (input) => [(0, exports.accountId32ToHex)(input)]);
