// @dedot/utils/cjs/hash/hashers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HASHERS = exports.identity = exports.twox64Concat = exports.twox256 = exports.twox128 = exports.blake2_128Concat = exports.blake2_256 = exports.blake2_128 = void 0;
const concat_js_1 = require("../concat.js");
const blake2_js_1 = require("./blake2.js");
const xxhash_js_1 = require("./xxhash.js");
const blake2_128 = (data) => (0, blake2_js_1.blake2AsU8a)(data, 128);
exports.blake2_128 = blake2_128;
const blake2_256 = (data) => (0, blake2_js_1.blake2AsU8a)(data, 256);
exports.blake2_256 = blake2_256;
const blake2_128Concat = (data) => (0, concat_js_1.concatU8a)((0, blake2_js_1.blake2AsU8a)(data, 128), data);
exports.blake2_128Concat = blake2_128Concat;
const twox128 = (data) => (0, xxhash_js_1.xxhashAsU8a)(data, 128);
exports.twox128 = twox128;
const twox256 = (data) => (0, xxhash_js_1.xxhashAsU8a)(data, 256);
exports.twox256 = twox256;
const twox64Concat = (data) => (0, concat_js_1.concatU8a)((0, xxhash_js_1.xxhashAsU8a)(data, 64), data);
exports.twox64Concat = twox64Concat;
const identity = (data) => data;
exports.identity = identity;
exports.HASHERS = {
    blake2_128: exports.blake2_128,
    blake2_256: exports.blake2_256,
    blake2_128Concat: exports.blake2_128Concat,
    twox128: exports.twox128,
    twox256: exports.twox256,
    twox64Concat: exports.twox64Concat,
    identity: exports.identity,
};
