// @dedot/shape/cjs/extension/str.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FixedStr = void 0;
const deshape_js_1 = require("../deshape.js");
const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();
function FixedStr(lengthInBytes) {
    return (0, deshape_js_1.createShape)({
        metadata: (0, deshape_js_1.metadata)('$FixedStr', FixedStr, lengthInBytes),
        staticSize: lengthInBytes,
        subEncode(buffer, value) {
            buffer.insertArray(textEncoder.encode(value));
        },
        subDecode(buffer) {
            if (buffer.array.length < buffer.index + lengthInBytes) {
                throw new deshape_js_1.ShapeDecodeError(this, buffer, 'Attempting to `str`-decode beyond bounds of input bytes');
            }
            const slice = buffer.array.subarray(buffer.index, buffer.index + lengthInBytes);
            buffer.index += lengthInBytes;
            return textDecoder.decode(slice);
        },
        subAssert(assert) {
            assert.typeof(this, 'string');
        },
    });
}
exports.FixedStr = FixedStr;
