// @dedot/api/cjs/extrinsic/extensions/known/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.knownSignedExtensions = void 0;
const ChargeAssetTxPayment_js_1 = require("./ChargeAssetTxPayment.js");
const ChargeTransactionPayment_js_1 = require("./ChargeTransactionPayment.js");
const CheckGenesis_js_1 = require("./CheckGenesis.js");
const CheckMetadataHash_js_1 = require("./CheckMetadataHash.js");
const CheckMortality_js_1 = require("./CheckMortality.js");
const CheckNonZeroSender_js_1 = require("./CheckNonZeroSender.js");
const CheckNonce_js_1 = require("./CheckNonce.js");
const CheckSpecVersion_js_1 = require("./CheckSpecVersion.js");
const CheckTxVersion_js_1 = require("./CheckTxVersion.js");
const CheckWeight_js_1 = require("./CheckWeight.js");
const PrevalidateAttests_js_1 = require("./PrevalidateAttests.js");
const StorageWeightReclaim_js_1 = require("./StorageWeightReclaim.js");
exports.knownSignedExtensions = {
    CheckNonZeroSender: CheckNonZeroSender_js_1.CheckNonZeroSender,
    CheckSpecVersion: CheckSpecVersion_js_1.CheckSpecVersion,
    CheckTxVersion: CheckTxVersion_js_1.CheckTxVersion,
    CheckGenesis: CheckGenesis_js_1.CheckGenesis,
    CheckMortality: CheckMortality_js_1.CheckMortality,
    CheckNonce: CheckNonce_js_1.CheckNonce,
    CheckWeight: CheckWeight_js_1.CheckWeight,
    ChargeTransactionPayment: ChargeTransactionPayment_js_1.ChargeTransactionPayment,
    PrevalidateAttests: PrevalidateAttests_js_1.PrevalidateAttests,
    ChargeAssetTxPayment: ChargeAssetTxPayment_js_1.ChargeAssetTxPayment,
    CheckMetadataHash: CheckMetadataHash_js_1.CheckMetadataHash,
    StorageWeightReclaim: StorageWeightReclaim_js_1.StorageWeightReclaim,
};
