// deshape/script/shapes/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constant = void 0;
const mod_js_1 = require("../common/mod.js");
function constant(value, c) {
    const pattern = c && (c instanceof Uint8Array ? c : c.encode(value));
    return (0, mod_js_1.createShape)({
        metadata: (0, mod_js_1.metadata)("$.constant", (constant), value, ...pattern ? [pattern] : []),
        // We could set `staticSize` to `pattern.length`, but in this case it will
        // usually more efficient to insert `pattern` dynamically, rather than
        // manually copy the bytes.
        staticSize: 0,
        subEncode(buffer) {
            if (pattern) {
                buffer.insertArray(pattern);
            }
        },
        subDecode(buffer) {
            if (pattern) {
                const got = buffer.array.subarray(buffer.index, buffer.index += pattern.length);
                for (let i = 0; i < pattern.length; i++) {
                    if (pattern[i] !== got[i]) {
                        throw new mod_js_1.ShapeDecodeError(this, buffer, `Invalid pattern; expected ${hex(pattern)}, got ${hex(got)}`);
                    }
                }
            }
            return value;
        },
        subAssert(assert) {
            assert.equals(this, value);
        },
    });
}
exports.constant = constant;
function hex(pattern) {
    let str = "0x";
    for (let i = 0; i < pattern.length; i++) {
        str += pattern[i].toString(16).padStart(2, "0");
    }
    return str;
}
