// deshape/script/shapes/lenPrefixed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lenPrefixed = void 0;
const mod_js_1 = require("../common/mod.js");
const compact_js_1 = require("./compact.js");
const int_js_1 = require("./int.js");
const compactU32 = (0, compact_js_1.compact)(int_js_1.u32);
function lenPrefixed($inner) {
    return (0, mod_js_1.createShape)({
        metadata: (0, mod_js_1.metadata)("$.lenPrefixed", lenPrefixed, $inner),
        staticSize: compactU32.staticSize + $inner.staticSize,
        subEncode(buffer, extrinsic) {
            const lengthCursor = buffer.createCursor(compactU32.staticSize);
            const contentCursor = buffer.createCursor($inner.staticSize);
            $inner.subEncode(contentCursor, extrinsic);
            buffer.waitForBuffer(contentCursor, () => {
                const length = contentCursor.finishedSize + contentCursor.index;
                compactU32.subEncode(lengthCursor, length);
                lengthCursor.close();
                contentCursor.close();
            });
        },
        subDecode(buffer) {
            const length = compactU32.subDecode(buffer);
            return $inner.decode(buffer.array.subarray(buffer.index, buffer.index += length));
        },
        subAssert(assert) {
            $inner.subAssert(assert);
        },
    });
}
exports.lenPrefixed = lenPrefixed;
