// deshape/script/shapes/str.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.str = void 0;
const mod_js_1 = require("../common/mod.js");
const compact_js_1 = require("./compact.js");
const int_js_1 = require("./int.js");
const compactU32 = (0, compact_js_1.compact)(int_js_1.u32);
const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();
exports.str = (0, mod_js_1.createShape)({
    metadata: (0, mod_js_1.metadata)("$.str"),
    staticSize: compactU32.staticSize,
    subEncode(buffer, value) {
        const array = textEncoder.encode(value);
        compactU32.subEncode(buffer, array.length);
        buffer.insertArray(array);
    },
    subDecode(buffer) {
        const len = compactU32.subDecode(buffer);
        if (buffer.array.length < buffer.index + len) {
            throw new mod_js_1.ShapeDecodeError(this, buffer, "Attempting to `str`-decode beyond bounds of input bytes");
        }
        const slice = buffer.array.subarray(buffer.index, buffer.index + len);
        buffer.index += len;
        return textDecoder.decode(slice);
    },
    subAssert(assert) {
        assert.typeof(this, "string");
    },
});
