// deshape/script/shapes/tuple.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tuple = void 0;
const mod_js_1 = require("../common/mod.js");
function tuple(...shapes) {
    return (0, mod_js_1.createShape)({
        metadata: (0, mod_js_1.metadata)("$.tuple", (tuple), ...shapes),
        staticSize: shapes.map((x) => x.staticSize).reduce((a, b) => a + b, 0),
        subEncode(buffer, value) {
            for (let i = 0; i < shapes.length; i++) {
                shapes[i].subEncode(buffer, value[i]);
            }
        },
        subDecode(buffer) {
            const value = Array(shapes.length);
            for (let i = 0; i < shapes.length; i++) {
                value[i] = shapes[i].subDecode(buffer);
            }
            return value;
        },
        subAssert(assert) {
            assert.instanceof(this, Array);
            assert.key(this, "length").equals(this, shapes.length);
            for (let i = 0; i < shapes.length; i++) {
                shapes[i].subAssert(assert.key(this, i));
            }
        },
    });
}
exports.tuple = tuple;
