// @dedot/api/cjs/executor/Executor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Executor = void 0;
const utils_1 = require("@dedot/utils");
/**
 * @name Executor
 * @description Execution abstraction for a specific action
 */
class Executor {
    client;
    #atBlockHash;
    constructor(client, atBlockHash) {
        this.client = client;
        this.#atBlockHash = atBlockHash;
    }
    get atBlockHash() {
        return this.#atBlockHash || this.client.atBlockHash;
    }
    get registry() {
        return this.client.registry;
    }
    get metadata() {
        return this.registry.metadata;
    }
    getPallet(name) {
        const targetPallet = this.metadata.pallets.find((p) => (0, utils_1.stringCamelCase)(p.name) === name);
        (0, utils_1.assert)(targetPallet, new utils_1.UnknownApiError(`Pallet not found: ${name}`));
        return targetPallet;
    }
    execute(...paths) {
        try {
            return this.doExecute(...paths);
        }
        catch (e) {
            if (!this.client.options?.throwOnUnknownApi && e instanceof utils_1.UnknownApiError) {
                return undefined;
            }
            throw e;
        }
    }
}
exports.Executor = Executor;
