// @dedot/api/cjs/extrinsic/extensions/SignedExtension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignedExtension = void 0;
const utils_1 = require("@dedot/utils");
class SignedExtension {
    client;
    options;
    data;
    additionalSigned;
    constructor(client, options) {
        this.client = client;
        this.options = options;
        this.data = {};
        this.additionalSigned = [];
    }
    async init() {
        // TODO implement this method
    }
    get identifier() {
        return this.signedExtensionDef.ident;
    }
    get $Data() {
        return (0, utils_1.ensurePresence)(this.registry.findCodec(this.signedExtensionDef.typeId));
    }
    get $AdditionalSigned() {
        return (0, utils_1.ensurePresence)(this.registry.findCodec(this.signedExtensionDef.additionalSigned));
    }
    get registry() {
        return this.client.registry;
    }
    get signedExtensionDef() {
        return (0, utils_1.ensurePresence)(this.options.def);
    }
    get payloadOptions() {
        return this.options?.payloadOptions || {};
    }
    toPayload(...args) {
        return {};
    }
}
exports.SignedExtension = SignedExtension;
