// @dedot/runtime-specs/cjs/core.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Core = void 0;
const codecs_1 = require("@dedot/codecs");
const $ = __importStar(require("@dedot/shape"));
/**
 * Ref: https://github.com/paritytech/polkadot-sdk/blob/eaf1bc5633ebbacce97e4f167ebe1d0d268c4b24/substrate/primitives/api/src/lib.rs#L799-L809
 */
exports.Core = [
    {
        methods: {
            version: {
                docs: 'Returns the version of the runtime.',
                params: [],
                type: 'RuntimeVersion',
                codec: codecs_1.$RuntimeVersion,
            },
            executeBlock: {
                docs: 'Execute the given block.',
                params: [
                    {
                        name: 'block',
                        type: 'Block',
                        codec: codecs_1.$Block,
                    },
                ],
                type: '[]',
                codec: $.Tuple(),
            },
            // Renamed at v2 (initialise_block)
            initializeBlock: {
                docs: 'Initialize a block with the given header.',
                params: [
                    {
                        name: 'header',
                        type: 'Header',
                        codec: codecs_1.$Header,
                    },
                ],
                type: '[]',
                codec: $.Tuple(),
            },
        },
        version: 4,
    },
];
