// @dedot/runtime-specs/cjs/genesis-builder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisBuilder = void 0;
const codecs_1 = require("@dedot/codecs");
const $ = __importStar(require("@dedot/shape"));
/**
 * Ref: https://github.com/paritytech/polkadot-sdk/blob/9a5d9a586e5a1c27bf29d3727b5edc89c59f0398/substrate/primitives/genesis-builder/src/lib.rs#L35-L45
 */
exports.GenesisBuilder = [
    {
        methods: {
            createDefaultConfig: {
                docs: [
                    'Creates the default `GenesisConfig` and returns it as a JSON blob.',
                    '\n',
                    'This function instantiates the default `GenesisConfig` struct for the runtime and serializes it into a JSON',
                    'blob. It returns a `Vec<u8>` containing the JSON representation of the default `GenesisConfig`.',
                ],
                params: [],
                type: 'Bytes',
                codec: codecs_1.$Bytes,
            },
            buildConfig: {
                docs: [
                    'Build `GenesisConfig` from a JSON blob not using any defaults and store it in the storage.',
                    '\n',
                    'This function deserializes the full `GenesisConfig` from the given JSON blob and puts it into the storage.',
                    'If the provided JSON blob is incorrect or incomplete or the deserialization fails, an error is returned.',
                    'It is recommended to log any errors encountered during the process.',
                    '\n',
                    'Please note that provided json blob must contain all `GenesisConfig` fields, no defaults will be used.',
                ],
                params: [
                    {
                        name: 'json',
                        type: 'Bytes',
                        codec: codecs_1.$Bytes,
                    },
                ],
                type: 'Result<[], Text>',
                codec: $.Result($.Tuple(), codecs_1.$Text),
            },
        },
        version: 1,
    },
];
