// @dedot/runtime-specs/cjs/mmr.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MmrApi = void 0;
const codecs_1 = require("@dedot/codecs");
const $ = __importStar(require("@dedot/shape"));
/**
 * Ref: https://github.com/paritytech/polkadot-sdk/blob/0e49ed72aa365475e30069a5c30e251a009fdacf/substrate/primitives/merkle-mountain-range/src/lib.rs#L423-L426
 */
exports.MmrApi = [
    {
        methods: {
            mmrRoot: {
                docs: 'Return the on-chain MMR root hash.',
                params: [],
                type: 'Result<Hash, MmrError>',
                codec: $.Result(codecs_1.$Hash, codecs_1.$MmrError),
            },
            mmrLeafCount: {
                docs: 'Return the number of MMR blocks in the chain.',
                params: [],
                type: 'Result<LeafIndex, MmrError>',
                codec: $.Result(codecs_1.$LeafIndex, codecs_1.$MmrError),
            },
            generateProof: {
                docs: [
                    'Generate MMR proof for a series of block numbers. If `best_known_block_number = Some(n)`,',
                    'use historical MMR state at given block height `n`. Else, use current MMR state.',
                ],
                params: [
                    {
                        name: 'blockNumbers',
                        type: 'Array<BlockNumber>',
                        codec: $.Array(codecs_1.$BlockNumber),
                    },
                    {
                        name: 'bestKnownBlockNumber',
                        type: 'Option<BlockNumber>',
                        codec: $.Option(codecs_1.$BlockNumber),
                    },
                ],
                type: 'Result<GeneratedMmrProofResult, MmrError>',
                codec: $.Result(codecs_1.$GeneratedMmrProofResult, codecs_1.$MmrError),
            },
            verifyProof: {
                docs: [
                    'Verify MMR proof against on-chain MMR for a batch of leaves.',
                    '\n',
                    'Note this function will use on-chain MMR root hash and check if the proof matches the hash.',
                    'Note, the leaves should be sorted such that corresponding leaves and leaf indices have the',
                    'same position in both the `leaves` vector and the `leaf_indices` vector contained in the [Proof]',
                ],
                params: [
                    {
                        name: 'leaves',
                        type: 'Array<MmrEncodableOpaqueLeaf>',
                        codec: $.Array(codecs_1.$MmrEncodableOpaqueLeaf),
                    },
                    {
                        name: 'proof',
                        type: 'MmrBatchProof',
                        codec: codecs_1.$MmrBatchProof,
                    },
                ],
                type: 'Result<[], MmrError>',
                codec: $.Result($.Tuple(), codecs_1.$MmrError),
            },
            verifyProofStateless: {
                docs: [
                    'Verify MMR proof against given root hash for a batch of leaves.',
                    '\n',
                    'Note this function does not require any on-chain storage - the',
                    'proof is verified against given MMR root hash.',
                    '\n',
                    'Note, the leaves should be sorted such that corresponding leaves and leaf indices have the',
                    'same position in both the `leaves` vector and the `leaf_indices` vector contained in the [Proof]',
                ],
                params: [
                    {
                        name: 'root',
                        type: 'Hash',
                        codec: codecs_1.$Hash,
                    },
                    {
                        name: 'leaves',
                        type: 'Array<MmrEncodableOpaqueLeaf>',
                        codec: $.Array(codecs_1.$MmrEncodableOpaqueLeaf),
                    },
                    {
                        name: 'proof',
                        type: 'MmrBatchProof',
                        codec: codecs_1.$MmrBatchProof,
                    },
                ],
                type: 'Result<[], MmrError>',
                codec: $.Result($.Tuple(), codecs_1.$MmrError),
            },
        },
        version: 2,
    },
];
