// @dedot/runtime-specs/cjs/nomination-pools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NominationPoolsApi = void 0;
const codecs_1 = require("@dedot/codecs");
/**
 * Ref: https://github.com/paritytech/polkadot-sdk/blob/bc53b9a03a742f8b658806a01a7bf853cb9a86cd/substrate/frame/nomination-pools/runtime-api/src/lib.rs#L26-L41
 */
exports.NominationPoolsApi = [
    {
        methods: {
            pendingRewards: {
                docs: 'Returns the pending rewards for the member that the AccountId was given for.',
                params: [
                    {
                        name: 'who',
                        type: 'AccountId32',
                        codec: codecs_1.$AccountId32,
                    },
                ],
                type: 'Balance',
                codec: codecs_1.$Balance,
            },
            pointsToBalance: {
                docs: 'Returns the equivalent balance of `points` for a given pool.',
                params: [
                    {
                        name: 'poolId',
                        type: 'NpPoolId',
                        codec: codecs_1.$NpPoolId,
                    },
                    {
                        name: 'points',
                        type: 'Balance',
                        codec: codecs_1.$Balance,
                    },
                ],
                type: 'Balance',
                codec: codecs_1.$Balance,
            },
            balanceToPoints: {
                docs: 'Returns the equivalent points of `new_funds` for a given pool.',
                params: [
                    {
                        name: 'poolId',
                        type: 'NpPoolId',
                        codec: codecs_1.$NpPoolId,
                    },
                    {
                        name: 'newFunds',
                        type: 'Balance',
                        codec: codecs_1.$Balance,
                    },
                ],
                type: 'Balance',
                codec: codecs_1.$Balance,
            },
        },
        version: 1,
    },
];
