// @dedot/runtime-specs/cjs/transaction-pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaggedTransactionQueue = void 0;
const codecs_1 = require("@dedot/codecs");
/**
 * Ref: https://github.com/paritytech/polkadot-sdk/blob/bc53b9a03a742f8b658806a01a7bf853cb9a86cd/substrate/primitives/transaction-pool/src/runtime_api.rs#L25-L30
 */
exports.TaggedTransactionQueue = [
    {
        methods: {
            validateTransaction: {
                docs: 'Validate the transaction.',
                params: [
                    {
                        name: 'source',
                        type: 'TransactionSource',
                        codec: codecs_1.$TransactionSource,
                    },
                    {
                        name: 'tx',
                        type: 'OpaqueExtrinsic',
                        codec: codecs_1.$OpaqueExtrinsic,
                    },
                    {
                        name: 'blockHash',
                        type: 'BlockHash',
                        codec: codecs_1.$BlockHash,
                    },
                ],
                type: 'TransactionValidity',
                codec: codecs_1.$TransactionValidity,
            },
        },
        version: 3,
    },
    {
        methods: {
            validateTransaction: {
                docs: 'Validate the transaction.',
                params: [
                    {
                        name: 'source',
                        type: 'TransactionSource',
                        codec: codecs_1.$TransactionSource,
                    },
                    {
                        name: 'tx',
                        type: 'OpaqueExtrinsic',
                        codec: codecs_1.$OpaqueExtrinsic,
                    },
                ],
                type: 'TransactionValidity',
                codec: codecs_1.$TransactionValidity,
            },
        },
        version: 2,
    },
    {
        methods: {
            validateTransaction: {
                docs: 'Validate the transaction.',
                params: [
                    {
                        name: 'tx',
                        type: 'OpaqueExtrinsic',
                        codec: codecs_1.$OpaqueExtrinsic,
                    },
                ],
                type: 'TransactionValidity',
                codec: codecs_1.$TransactionValidity,
            },
        },
        version: 1,
    },
];
