// @dedot/codecs/cjs/codecs/metadata/v14.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$MetadataV14 = exports.$ExtrinsicDefV14 = exports.$SignedExtensionDefV14 = exports.$PalletDefV14 = exports.$ConstantDefV14 = exports.$StorageEntryV14 = exports.$Hasher = void 0;
const $ = __importStar(require("@dedot/shape"));
const scale_info_js_1 = require("./scale_info.js");
exports.$Hasher = $.FlatEnum([
    'blake2_128',
    'blake2_256',
    'blake2_128Concat',
    'twox128',
    'twox256',
    'twox64Concat',
    'identity',
]);
exports.$StorageEntryV14 = $.Struct({
    name: $.str,
    modifier: $.FlatEnum(['Optional', 'Default']),
    storageType: $.Enum({
        Plain: $.Struct({ valueTypeId: scale_info_js_1.$TypeId }),
        Map: $.Struct({
            hashers: $.Vec(exports.$Hasher),
            keyTypeId: scale_info_js_1.$TypeId,
            valueTypeId: scale_info_js_1.$TypeId,
        }),
    }),
    default: $.PrefixedHex,
    docs: $.Vec($.str),
});
exports.$ConstantDefV14 = $.Struct({
    name: $.str,
    typeId: scale_info_js_1.$TypeId,
    value: $.PrefixedHex,
    docs: $.Vec($.str),
});
exports.$PalletDefV14 = $.Struct({
    name: $.str,
    storage: $.Option($.Struct({
        prefix: $.str,
        entries: $.Vec(exports.$StorageEntryV14),
    })),
    calls: $.Option(scale_info_js_1.$TypeId),
    event: $.Option(scale_info_js_1.$TypeId),
    constants: $.Vec(exports.$ConstantDefV14),
    error: $.Option(scale_info_js_1.$TypeId),
    index: $.u8,
});
exports.$SignedExtensionDefV14 = $.Struct({
    ident: $.str,
    typeId: scale_info_js_1.$TypeId,
    additionalSigned: scale_info_js_1.$TypeId,
});
exports.$ExtrinsicDefV14 = $.Struct({
    typeId: scale_info_js_1.$TypeId,
    version: $.u8,
    signedExtensions: $.Vec(exports.$SignedExtensionDefV14),
});
/**
 * Ref: https://github.com/paritytech/frame-metadata/blob/a07b2451b82809501fd797691046c1164f7e8840/frame-metadata/src/v14.rs#L45-L54
 */
exports.$MetadataV14 = $.Struct({
    types: $.Vec(scale_info_js_1.$PortableType),
    pallets: $.Vec(exports.$PalletDefV14),
    extrinsic: exports.$ExtrinsicDefV14,
    runtimeType: scale_info_js_1.$TypeId,
});
