// @dedot/codecs/cjs/codecs/metadata/v15.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$MetadataV15 = exports.$RuntimeApiDefV15 = exports.$RuntimeApiMethodDefV15 = exports.$RuntimeApiMethodParamDefV15 = exports.$ExtrinsicDefV15 = exports.$SignedExtensionDefV15 = exports.$PalletDefV15 = exports.$StorageEntryV15 = exports.$ConstantDefV15 = void 0;
const $ = __importStar(require("@dedot/shape"));
const scale_info_js_1 = require("./scale_info.js");
const v14_js_1 = require("./v14.js");
exports.$ConstantDefV15 = v14_js_1.$ConstantDefV14;
exports.$StorageEntryV15 = v14_js_1.$StorageEntryV14;
exports.$PalletDefV15 = $.Struct({
    name: $.str,
    storage: $.Option($.Struct({
        prefix: $.str,
        entries: $.Vec(exports.$StorageEntryV15),
    })),
    calls: $.Option(scale_info_js_1.$TypeId),
    event: $.Option(scale_info_js_1.$TypeId),
    constants: $.Vec(exports.$ConstantDefV15),
    error: $.Option(scale_info_js_1.$TypeId),
    index: $.u8,
    docs: $.Vec($.str),
});
exports.$SignedExtensionDefV15 = v14_js_1.$SignedExtensionDefV14;
exports.$ExtrinsicDefV15 = $.Struct({
    version: $.u8,
    addressTypeId: scale_info_js_1.$TypeId,
    callTypeId: scale_info_js_1.$TypeId,
    signatureTypeId: scale_info_js_1.$TypeId,
    extraTypeId: scale_info_js_1.$TypeId,
    signedExtensions: $.Vec(exports.$SignedExtensionDefV15),
});
exports.$RuntimeApiMethodParamDefV15 = $.Struct({
    name: $.str,
    typeId: scale_info_js_1.$TypeId,
});
exports.$RuntimeApiMethodDefV15 = $.Struct({
    name: $.str,
    inputs: $.Vec(exports.$RuntimeApiMethodParamDefV15),
    output: scale_info_js_1.$TypeId,
    docs: $.Vec($.str),
});
exports.$RuntimeApiDefV15 = $.Struct({
    name: $.str,
    methods: $.Vec(exports.$RuntimeApiMethodDefV15),
    docs: $.Vec($.str),
});
/**
 * Ref: https://github.com/paritytech/frame-metadata/blob/a07b2451b82809501fd797691046c1164f7e8840/frame-metadata/src/v15.rs#L48-L63
 */
exports.$MetadataV15 = $.Struct({
    types: $.Vec(scale_info_js_1.$PortableType),
    pallets: $.Vec(exports.$PalletDefV15),
    extrinsic: exports.$ExtrinsicDefV15,
    runtimeType: scale_info_js_1.$TypeId,
    apis: $.Vec(exports.$RuntimeApiDefV15),
    outerEnums: $.Struct({
        callEnumTypeId: scale_info_js_1.$TypeId,
        eventEnumTypeId: scale_info_js_1.$TypeId,
        errorEnumTypeId: scale_info_js_1.$TypeId,
    }),
    custom: $.Struct({
        map: $.map($.str, $.Struct({
            typeId: scale_info_js_1.$TypeId,
            value: $.PrefixedHex,
        })),
    }),
});
