// @dedot/codecs/cjs/codecs/known/arithmetic.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$FixedU128 = exports.$FixedI128 = exports.$FixedU64 = exports.$FixedI64 = exports.$Perquintill = exports.$Perbill = exports.$Permill = exports.$PerU16 = exports.$Percent = void 0;
const $ = __importStar(require("@dedot/shape"));
exports.$Percent = $.u8; // 0 -> 100u8
exports.$PerU16 = $.u16; // 0 -> 65535_u16
exports.$Permill = $.u32; // 0 -> 1_000_000u32
exports.$Perbill = $.u32; // 0 -> 1_000_000_000u32
exports.$Perquintill = $.u64; // 0 -> 1_000_000_000_000_000_000u64
exports.$FixedI64 = $.i64;
exports.$FixedU64 = $.u64;
exports.$FixedI128 = $.i128;
exports.$FixedU128 = $.u128;
