// @dedot/codecs/cjs/codecs/known/common.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$Bytes = exports.$RawBytes = exports.$StorageData = exports.$PrefixedStorageKey = exports.$StorageKey = exports.$Text = void 0;
const $ = __importStar(require("@dedot/shape"));
const utils_1 = require("@dedot/utils");
exports.$Text = $.str;
exports.$StorageKey = $.RawHex;
exports.$PrefixedStorageKey = $.RawHex;
exports.$StorageData = $.RawHex;
exports.$RawBytes = $.transform({
    $base: $.RawHex,
    encode: (value) => ((0, utils_1.isU8a)(value) ? (0, utils_1.u8aToHex)(value) : value),
    decode: (value) => value,
});
exports.$Bytes = $.transform({
    $base: $.PrefixedHex,
    encode: (value) => (0, utils_1.toHex)(value),
    decode: (value) => value,
});
