// @dedot/codecs/cjs/codecs/known/primitives.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$H768 = exports.$H512 = exports.$H384 = exports.$H256 = exports.$H160 = exports.$H128 = exports.$U256 = exports.$U128 = void 0;
const $ = __importStar(require("@dedot/shape"));
exports.$U128 = $.u128;
exports.$U256 = $.u256;
// TODO to implement!
// export const $U512 = ?!?
exports.$H128 = $.FixedHex(16);
exports.$H160 = $.FixedHex(20);
exports.$H256 = $.FixedHex(32);
exports.$H384 = $.FixedHex(48);
exports.$H512 = $.FixedHex(64);
exports.$H768 = $.FixedHex(96);
